File: backend/server-cors-example.ts
/**
 * Express.js Backend CORS Configuration for Codespaces
 * 
 * This setup allows your Express backend to accept requests from both:
 * - Local development: http://localhost:5173
 * - GitHub Codespaces: https://<codespace-name>-5173.app.github.dev
 */

import cors from 'cors';
import express from 'express';

const app = express();

/**
 * Determine if running in GitHub Codespaces
 */
const isCodespaces = (): boolean => {
  return !!(process.env.CODESPACE_NAME);
};

/**
 * Get allowed origins based on environment
 */
const getAllowedOrigins = (): string[] => {
  const codespaceName = process.env.CODESPACE_NAME;
  const frontendPort = process.env.VITE_FRONTEND_PORT || '5173';

  if (isCodespaces() && codespaceName) {
    return [
      // Allow Codespaces URL
      `https://${codespaceName}-${frontendPort}.app.github.dev`,
      // Also allow localhost for local testing inside container
      `http://localhost:${frontendPort}`,
    ];
  }

  // Local development - allow localhost
  return [
    `http://localhost:${frontendPort}`,
    'http://127.0.0.1:3000', // Alternative local address
  ];
};

/**
 * CORS Configuration
 */
const corsOptions: cors.CorsOptions = {
  origin: (origin, callback) => {
    const allowedOrigins = getAllowedOrigins();

    // Allow requests with no origin (like mobile apps or server-to-server)
    if (!origin || allowedOrigins.includes(origin)) {
      callback(null, true);
    } else {
      callback(new Error('Not allowed by CORS'), false);
    }
  },

  credentials: true, // Allow cookies and authentication headers
  optionsSuccessStatus: 200, // For legacy browsers
  methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS', 'HEAD'],
  allowedHeaders: [
    'Content-Type',
    'Authorization',
    'X-Requested-With',
    'Accept',
    'Origin',
  ],
  exposedHeaders: ['X-Total-Count', 'X-Page-Number', 'X-Page-Size'],
  maxAge: 3600, // Cache CORS preflight responses for 1 hour
};

/**
 * Apply CORS middleware
 */
app.use(cors(corsOptions));

/**
 * Handle preflight requests explicitly (important for some browsers)
 */
app.options('*', cors(corsOptions));

/**
 * Body parsing middleware
 */
app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ limit: '10mb', extended: true }));

/**
 * Logging middleware - useful for debugging Codespaces issues
 */
app.use((req, res, next) => {
  console.log(`[${new Date().toISOString()}] ${req.method} ${req.path}`);
  console.log('Origin:', req.get('origin'));
  console.log('Host:', req.get('host'));
  next();
});

/**
 * Health check endpoint
 */
app.get('/api/health', (req, res) => {
  res.json({
    status: 'ok',
    environment: isCodespaces() ? 'codespaces' : 'local',
    codespace: process.env.CODESPACE_NAME || 'N/A',
    timestamp: new Date().toISOString(),
  });
});

/**
 * Example API route
 */
app.post('/api/data', (req, res) => {
  try {
    // Your endpoint logic here
    res.json({
      message: 'Data received successfully',
      data: req.body,
    });
  } catch (error) {
    console.error('Error:', error);
    res.status(500).json({
      error: 'Internal server error',
    });
  }
});

/**
 * Error handling middleware
 */
app.use(
  (
    err: any,
    req: express.Request,
    res: express.Response,
    next: express.NextFunction
  ) => {
    console.error('Error:', err);
    res.status(err.status || 500).json({
      error: err.message || 'Internal server error',
    });
  }
);

/**
 * Start server
 */
const PORT = process.env.PORT || 5000;
app.listen(PORT, () => {
  console.log(`\n🚀 Server running on port ${PORT}`);
  console.log(`📡 CORS enabled for origins:`, getAllowedOrigins());
  console.log(`🌍 Environment: ${isCodespaces() ? 'GitHub Codespaces' : 'Local'}`);
  if (isCodespaces()) {
    console.log(`📦 Codespace: ${process.env.CODESPACE_NAME}`);
  }
  console.log('');
});

export { app, getAllowedOrigins, isCodespaces };

File: backend/server-cors-example.py
"""
Flask Backend CORS Configuration for Codespaces

This setup allows your Flask backend to accept requests from both:
- Local development: http://localhost:5173
- GitHub Codespaces: https://<codespace-name>-5173.app.github.dev
"""

import os
from flask import Flask, jsonify, request
from flask_cors import CORS
from datetime import datetime

app = Flask(__name__)


def is_codespaces() -> bool:
    """Determine if running in GitHub Codespaces"""
    return bool(os.getenv('CODESPACE_NAME'))


def get_allowed_origins() -> list:
    """Get allowed origins based on environment"""
    codespace_name = os.getenv('CODESPACE_NAME')
    frontend_port = os.getenv('VITE_FRONTEND_PORT', '5173')

    if is_codespaces() and codespace_name:
        return [
            # Allow Codespaces URL
            f'https://{codespace_name}-{frontend_port}.app.github.dev',
            # Also allow localhost for local testing inside container
            f'http://localhost:{frontend_port}',
        ]

    # Local development - allow localhost
    return [
        f'http://localhost:{frontend_port}',
        'http://127.0.0.1:3000',  # Alternative local address
    ]


# CORS Configuration
CORS(
    app,
    resources={r'/api/*': {
        'origins': get_allowed_origins(),
        'methods': ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS', 'HEAD'],
        'allow_headers': [
            'Content-Type',
            'Authorization',
            'X-Requested-With',
            'Accept',
            'Origin',
        ],
        'expose_headers': [
            'X-Total-Count',
            'X-Page-Number',
            'X-Page-Size',
        ],
        'supports_credentials': True,
        'max_age': 3600,  # Cache CORS preflight for 1 hour
    }},
)


@app.before_request
def log_request():
    """Log incoming requests - useful for debugging Codespaces issues"""
    print(f'[{datetime.now().isoformat()}] {request.method} {request.path}')
    print(f'Origin: {request.headers.get("Origin")}')
    print(f'Host: {request.headers.get("Host")}')


@app.route('/api/health', methods=['GET'])
def health_check():
    """Health check endpoint"""
    return jsonify({
        'status': 'ok',
        'environment': 'codespaces' if is_codespaces() else 'local',
        'codespace': os.getenv('CODESPACE_NAME', 'N/A'),
        'timestamp': datetime.now().isoformat(),
    })


@app.route('/api/data', methods=['POST'])
def handle_data():
    """Example API route"""
    try:
        data = request.get_json()
        return jsonify({
            'message': 'Data received successfully',
            'data': data,
        }), 200
    except Exception as error:
        print(f'Error: {error}')
        return jsonify({
            'error': 'Internal server error',
        }), 500


@app.errorhandler(404)
def not_found(error):
    """Handle 404 errors"""
    return jsonify({
        'error': 'Not found',
    }), 404


@app.errorhandler(500)
def internal_error(error):
    """Handle 500 errors"""
    print(f'Error: {error}')
    return jsonify({
        'error': 'Internal server error',
    }), 500


if __name__ == '__main__':
    port = int(os.getenv('PORT', 5000))
    allowed_origins = get_allowed_origins()

    print(f'\n🚀 Server running on port {port}')
    print(f'📡 CORS enabled for origins: {allowed_origins}')
    print(f'🌍 Environment: {"GitHub Codespaces" if is_codespaces() else "Local"}')
    if is_codespaces():
        print(f'📦 Codespace: {os.getenv("CODESPACE_NAME")}')
    print('')

    # Run Flask app
    app.run(
        host='0.0.0.0',  # Allow connections from anywhere
        port=port,
        debug=os.getenv('FLASK_ENV') == 'development',
    )
